tool/tsh/kubectl.go

```diff
@@ -275,3 +313,178 @@ func getKubeClusterName(args []string, teleportClusterName string) (string, erro
+	kubectlArgs []string
+	// makeAndStartKubeLocalProxyFunc is a callback function to create and
+	// start a kube local proxy, when it is decided that a local proxy is
+	// required. Default to makeAndStartKubeLocalProxy. Can be set another
+	// function for testing.
+	makeAndStartKubeLocalProxyFunc func(*CLIConf, *clientcmdapi.Config, kubeconfig.LocalProxyClusters) (func(), string, error)
+	// onNewKubeconfigFuncs is list of callback functions that will receive the
+	// new kubeconfig location of the local proxy when the local proxy is
+	// successfully created.
+	onNewKubeconfigFuncs []func(string) error
+}
+
+type applyKubeLocalProxyOpts func(o *kubeLocalProxyOpts)
+
+func withKubectlArgs(args []string) applyKubeLocalProxyOpts {
+	return func(o *kubeLocalProxyOpts) {
+		o.kubectlArgs = args
+		o.onNewKubeconfigFuncs = append(o.onNewKubeconfigFuncs, func(newPath string) error {
+			overwriteKubeconfigFlagInArgs(args, newPath)
+			return nil
+		})
+	}
+}
+func withKubeconfigEnvUpdate() applyKubeLocalProxyOpts {
+	return func(o *kubeLocalProxyOpts) {
+		o.onNewKubeconfigFuncs = append(o.onNewKubeconfigFuncs, func(newPath string) error {
+			return trace.Wrap(os.Setenv(teleport.EnvKubeConfig, newPath))
+		})
+	}
+}
+
+func newKubeLocalProxyOpts(applyOpts ...applyKubeLocalProxyOpts) kubeLocalProxyOpts {
+	opts := kubeLocalProxyOpts{
+		makeAndStartKubeLocalProxyFunc: makeAndStartKubeLocalProxy,
+	}
+	for _, applyOpt := range applyOpts {
+		applyOpt(&opts)
+	}
+	return opts
+}
+
+// maybeStartKubeLocalProxy starts a kube local proxy if local proxy is
+// required, and calls provided onNewKubeconfigFuncs to update the kubeconfig
+// path. Called by `tsh kubectl` and `tsh kube exec`.
+func maybeStartKubeLocalProxy(cf *CLIConf, applyOpts ...applyKubeLocalProxyOpts) (func(), error) {
+	opts := newKubeLocalProxyOpts(applyOpts...)
+
+	config, clusters, useLocalProxy := shouldUseKubeLocalProxy(cf, opts.kubectlArgs)
+	if !useLocalProxy {
+		return func() {}, nil
+	}
+
+	closeFn, newKubeconfigLocation, err := opts.makeAndStartKubeLocalProxyFunc(cf, config, clusters)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	for _, onNewLocation := range opts.onNewKubeconfigFuncs {
+		if err := onNewLocation(newKubeconfigLocation); err != nil {
+			defer closeFn()
+			return nil, trace.Wrap(err)
+		}
+	}
+
+	return closeFn, nil
+}
+
+// makeAndStartKubeLocalProxy is a helper to create a kube local proxy and
+// start it in a goroutine. If successful, a closeFn and the generated
+// kubeconfig location are returned.
+func makeAndStartKubeLocalProxy(cf *CLIConf, config *clientcmdapi.Config, clusters kubeconfig.LocalProxyClusters) (func(), string, error) {
+	tc, err := makeClient(cf, true)
+	if err != nil {
+		return nil, "", trace.Wrap(err)
+	}
+
+	localProxy, err := makeKubeLocalProxy(cf, tc, clusters, config, cf.LocalProxyPort)
+	if err != nil {
+		return nil, "", trace.Wrap(err)
+	}
+
+	go localProxy.Start(cf.Context)
```

Comment:
I wonder how the SIGTERM/SIGINT signal and interactive command behaves when the local proxy is run in background like in case of `tsh kubectl exec --stdin --tty pod01  -- /bin/sh` and CTRL+C signal.